src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ContextAware.cs

```diff
@@ -0,0 +1,103 @@
+    {
+        private (DataContract? strong, WeakReference<DataContract>? weak)[] _contracts;
+        private ConditionalWeakTable<Type, DataContract> _keepAlive;
+
+        public int Length => _contracts.Length;
+
+        public ContextAwareDataContractIndex(int size)
+        {
+            _contracts = new (DataContract?, WeakReference<DataContract>?)[size];
+            _keepAlive = new ConditionalWeakTable<Type, DataContract>();
+        }
+
+        public DataContract? GetItem(int index) => _contracts[index].strong ?? (_contracts[index].weak?.TryGetTarget(out DataContract? ret) == true ? ret : null);
+
+        public void SetItem(int index, DataContract dataContract)
+        {
+            // Check for unloadability to decide how to store the value
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(dataContract.UnderlyingType.Assembly);
+            if (alc == null || !alc.IsCollectible)
+            {
+                _contracts[index].strong = dataContract;
+            }
+            else
+            {
+                _contracts[index].weak = new WeakReference<DataContract>(dataContract);
+                _keepAlive.Add(dataContract.UnderlyingType, dataContract);
+            }
+        }
+
+        public void Resize(int newSize)
+        {
+            Array.Resize<(DataContract?, WeakReference<DataContract>?)>(ref _contracts, newSize);
+        }
+    }
+
+    internal sealed class ContextAwareDictionary<TKey, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] TValue>
+        where TKey : Type
+        where TValue : class?
+    {
+        private readonly ConcurrentDictionary<TKey, TValue> _fastDictionary = new();
+        private readonly ConditionalWeakTable<TKey, TValue> _collectibleTable = new();
+
+
+        internal TValue GetOrAdd(TKey t, Func<TKey, TValue> f)
+        {
+            TValue? ret;
+
+            // The fast and most common default case
+            if (_fastDictionary.TryGetValue(t, out ret))
+                return ret;
+
+            // Common case for collectible contexts
+            if (_collectibleTable.TryGetValue(t, out ret))
+                return ret;
+
+            // Not found. Do the slower work of creating the value in the correct collection.
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(t.Assembly);
+
+            // Null and non-collectible load contexts use the default table
+            if (alc == null || !alc.IsCollectible)
+            {
+                // The create delegate here could be quite expensive. ConcurrentDictionary semantics would let use
+                // do this without a lock and not corrupt the dictionary, but the delegate still might be called multiple
+                // times. So we use a lock to ensure the delegate is only called once. Keep the lock off the hot path.
+                if (!_fastDictionary.TryGetValue(t, out ret))
+                {
```

Comment:
The TryGetValue should be inside the lock to avoid race conditions (which would be non-critical, but still something you want to avoid). Same for the collectibleTable.